File: check.py
import json
from datetime import datetime, timedelta
import urllib.request
import urllib.parse

def check_key(key):
    apiUrl = "https://api.openai.com"
    urlSubscription = f"{apiUrl}/v1/dashboard/billing/subscription"
    headers = {
        "Authorization": f"Bearer {key}",
        "Content-Type": "application/json",
    }

    req_subscription = urllib.request.Request(urlSubscription, headers=headers, method='GET')
    with urllib.request.urlopen(req_subscription) as response:
        if response.status != 200:
            return {"error": "您的账户已被封禁，请登录OpenAI进行查看。"}
        subscriptionData = json.loads(response.read().decode())
        totalAmount = subscriptionData['hard_limit_usd']

    now = datetime.now()
    startDate = (now - timedelta(days=90)).strftime('%Y-%m-%d')
    endDate = (now + timedelta(days=1)).strftime('%Y-%m-%d')
    subDate = now.replace(day=1).strftime('%Y-%m-%d')
    if totalAmount > 20:
        startDate = subDate

    urlUsage = f"{apiUrl}/v1/dashboard/billing/usage?start_date={startDate}&end_date={endDate}"
    req_usage = urllib.request.Request(urlUsage, headers=headers, method='GET')
    with urllib.request.urlopen(req_usage) as response:
        usageData = json.loads(response.read().decode())
        totalUsage = usageData['total_usage'] / 100
        remaining = totalAmount - totalUsage

    response_body = json.dumps({
        "total_amount": format(totalAmount, '.2f'),
        "total_used": format(totalUsage, '.2f'),
        "remaining": format(remaining, '.2f')
    })

    return json.loads(response_body)

result = check_key("sk-xxxxxxxx......")
print(result)

File: check.php
<?php
  $request_body = file_get_contents('php://input');
  $apikey = json_decode($request_body)->apikey;

  if (!$apikey) {
    http_response_code(400);
    echo "请提供 API KEY";
    exit();
  }

  $apiUrl = "https://api.openai.com";
  $urlSubscription = "{$apiUrl}/v1/dashboard/billing/subscription";
  $now = new DateTime();
  $startDate = new DateTime("-90 days");
  $endDate = (new DateTime())->add(new DateInterval("P1D"));
  $subDate = new DateTime();
  $subDate->setDate($now->format("Y"), $now->format("m"), 1);

  $headers = [
    "Authorization" => "Bearer {$apikey}",
    "Content-Type" => "application/json",
  ];

  $response = file_get_contents($urlSubscription, false, stream_context_create([
    "http" => [
      "method" => "GET",
      "header" => implode("\r\n", array_map(function($key, $value) {
        return "{$key}: {$value}";
      }, array_keys($headers), $headers))
    ]
  ]));

  if (!$response) {
    http_response_code(400);
    echo "您的账户已被封禁，请登录OpenAI进行查看。";
    exit();
  }

  $subscriptionData = json_decode($response);
  $totalAmount = $subscriptionData->hard_limit_usd;

  if ($totalAmount > 20) {
    $startDate = $subDate;
  }

  $urlUsage = "{$apiUrl}/v1/dashboard/billing/usage?start_date={$startDate->format("Y-m-d")}&end_date={$endDate->format("Y-m-d")}";

  $response = file_get_contents($urlUsage, false, stream_context_create([
    "http" => [
      "method" => "GET",
      "header" => implode("\r\n", array_map(function($key, $value) {
        return "{$key}: {$value}";
      }, array_keys($headers), $headers))
    ]
  ]));

  $usageData = json_decode($response);
  $totalUsage = $usageData->total_usage / 100;
  $remaining = $totalAmount - $totalUsage;

  $response_body = json_encode([
    "total_amount" => number_format($totalAmount, 2),
    "total_used" => number_format($totalUsage, 2),
    "remaining" => number_format($remaining, 2)
  ]);

  header("Content-Type: application/json");
  echo $response_body;
